ex4/VendingDP.Lib/Menu.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace VendingDP.Lib
+{
+    public class Menu : MenuLoader
+    {
+        private readonly ErrorHandler errorHandler = new ErrorHandler();
+        public Menu()
+        {
+            Validate();
+        }
+
+        public void Validate()
+        {
+            CheckDataCollections();
+            IEnumerable<string> wrongCategoryToppings = WrongCategoryToppings();
+            errorHandler.InvalidMenuItems(wrongCategoryToppings, "toppings");
+
+            IEnumerable<string> wrongTypeCombinations = WronglyWrittenPreparedCombinations();
+            errorHandler.InvalidMenuItems(wrongTypeCombinations, "prepared combinations");
+        }
+
+        private void CheckDataCollections()
+        {
+            errorHandler.IsDataCollectionEmpty(BasicProducts, "basic products");
+            errorHandler.IsDataCollectionEmpty(Toppings, "toppings");
+            errorHandler.IsDataCollectionEmpty(PreparedCombinations, "prepared combinations");
+        }
+
+        private IEnumerable<string> WrongCategoryToppings()
+        {
+            IEnumerable<string> validCategories = GetCategories();
+            IEnumerable<string> wrongCategoryToppings = Toppings
+                .Where(topping => !validCategories.Contains(topping.Category)).Select(topping => topping.Name);
+            return wrongCategoryToppings;
+        }
+
+        private IEnumerable<string> WronglyWrittenPreparedCombinations() => PreparedCombinations
+            .Where(rawPreparedCombination => !ValidateCombinatedProduct(rawPreparedCombination))
+            .Select(combination => combination.Name);
+
+        private bool ValidateCombinatedProduct(RawPreparedCombination combinatedProduct)
+        {
+            bool isFirstBasicProduct = BasicProducts.Select(basicProduct => basicProduct.Name).Contains(combinatedProduct.BasicProductName);
+            IEnumerable<string> toppingNames = Toppings.Select(topping => topping.Name);
+            bool areRestToppings = combinatedProduct.Toppings.All(toppingName => toppingNames.Contains(toppingName));
+            return isFirstBasicProduct && areRestToppings;
+        }
+
+        private Dictionary<string, Dictionary<string, List<Item>>> ItemsToCategories()
+        {
+            Dictionary<string, Dictionary<string, List<Item>>> itemsToCategories = new Dictionary<string, Dictionary<string, List<Item>>>();
+            IEnumerable<string> categories = GetCategories();
+            foreach (string category in categories)
+            {
+                itemsToCategories.Add(category, new Dictionary<string, List<Item>>()
+                {
+                    ["Basic Products"] = new List<Item>(),
```

Comment:
Instead of using a dict use a new class, dicts are the source of problems! What happens when a key is not in the dicr? it makes you check it and make the code much more complicated.

Reply:
The thing is I want to divide the data (basic products and toppings) into categories, so when I present them in the menu it will be by category. For this, the most comfortable way I see is to hold the data in a dictionary, where the keys are the category names and the values can be a class holding a collection of matching basic products and a collection of matching toppings. 
Is it fine? If not, what better way do you see that could help me match every category name to the items from this category (efficiently)?

Reply:
Yes (if I got it right). you are trying to create a map Category -> Item (name), and group by their key type (basic product, toppings). It sounds like linq..